### ISSUE TITLE
Improve performance of the Nearest Neighbors search

### ISSUE THREAD

**Issue by rth** [FreeDiscovery/FreeDiscovery#59]
While the current performance of Nearest Neighbors categorization in the LSI space remains reasonable, it is not very scalable to large training sets (>5k documents), as discussed in issue #58. 

There is a number of options in the [`NearestNeighborRanker`](https://github.com/FreeDiscovery/FreeDiscovery/blob/master/freediscovery/categorization.py#L104) class, inherited from scikit-learn's [`NearestNeighbors`](http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html) implementation including,
  * algorithm
  * leaf_size
  * radius

We are currently using the default values `algorithm="ball_tree", leaf_size=30, radius=1.0`, however that might not be optimal with the dataset we have, which is,

   * for the training set `X_train` of shape `[n_train, n_lsi]` (where `n_train ~ 1e3 - 5e3` and `n_lsi ~ 150`)
   * for the test set  `X` of shape `[n_samples, n_lsi]` (where `n_samples ~ 500e3` and `n_lsi ~ 150`). 

`X` is always L2 normalizes so it's values are close to 1.

I believe `ball_tree` algorithm is still the probably best here, but reducing `radius` or changing `leaf_size` might improve run times. 

Here is a sample  benchmark script, **nn_benchmark.py** 
```py
# -*- coding: utf-8 -*-

from __future__ import print_function

from time import time

import numpy as np
import scipy.linalg
from sklearn.neighbors import NearestNeighbors
from sklearn.preprocessing import normalize


n_samples = 20000
n_lsi = 150

np.random.seed(999999)

X = np.random.randn(n_samples, n_lsi)
normalize(X, norm='l2', copy=False)

print('Test set: {} docs, {} LSI dim'.format(n_samples, n_lsi))

for N_train in [1000, 2000]:
    mod = NearestNeighbors(n_neighbors=1, # always at 1 in FreeDiscovery
                          algorithm='ball_tree',
                          radius=1.0,
                          leaf_size=30,
                          n_jobs=1 # value of 1 best for the benchmark
                          )

    t0 = time()
    mod.fit(X[:N_train, :])
    t1 = time()
    mod.kneighbors(X, return_distance=True)
    t2 = time()

    print('Training set {} docs, t_train= {:.2f} s, t_test = {:.2f}s'.format(
                        N_train, t1-t0, t2-t1))
```

which can be run (on Linux) with,
```bash
$ source activate freediscovery-env
$ /usr/bin/time -v python nn_benchmark.py   
Test set: 20000 docs, 150 LSI dim
Training set 1000 docs, t_train= 0.01 s, t_test = 5.35s
Training set 2000 docs, t_train= 0.01 s, t_test = 11.02s
        [...]
        Maximum resident set size (kbytes): 81752
        [...]
```
where the `Maximum resident set size (kbytes)` is the RAM usage. Some parameters (e.g. `algorithm=brute`) might be occasionally faster but would use an unreasobable amounts of RAM.

If anyone is interested in investigating this, it would be quite useful...

**Comment by rth** [comment]
Actually, it looks like the brute force nearest neighbor search is most efficient for these dataset sizes (cf. https://github.com/scikit-learn/scikit-learn/issues/8213) also care must be taken not to use `n_jobs > 1` parameter, as that significantly slows down the calculation (up to x5-x8 on 4 CPU cores for 700k training / 1k test dataset) as discussed in https://github.com/scikit-learn/scikit-learn/issues/8216 .

**Comment by rth** [comment]
The way forward here could be taking an approximate NN algorithm from [this comparison](https://github.com/erikbern/ann-benchmarks) or possibly [faiss](https://github.com/facebookresearch/faiss)...
